src/OpenTelemetry/Logs/LogRecord.cs

```diff
@@ -213,22 +221,77 @@ public void ForEachScope<TState>(Action<LogRecordScope, TState> callback, TState
             return ref this.Data;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void ResetReferenceCount()
+        {
+            this.PoolReferenceCount = 1;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void AddReference()
+        {
+            Interlocked.Increment(ref this.PoolReferenceCount);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal int RemoveReference()
+        {
+            return Interlocked.Decrement(ref this.PoolReferenceCount);
+        }
+
+        internal void Buffer()
+        {
+            this.BufferLogStateValues();
+            this.BufferLogScopes();
+        }
+
+        internal LogRecord Copy()
+        {
+            this.Buffer();
+
+            return new()
+            {
+                Data = this.Data,
+                State = this.State,
+                StateValues = this.StateValues == null ? null : new List<KeyValuePair<string, object?>>(this.StateValues),
+                BufferedScopes = this.BufferedScopes == null ? null : new List<object?>(this.BufferedScopes),
+            };
+        }
+
+        /// <summary>
+        /// Buffers the state values attached to the log into a list so that
+        /// they can be safely processed after the log message lifecycle has
+        /// ended.
+        /// </summary>
+        private void BufferLogStateValues()
+        {
+            var stateValues = this.StateValues;
+            if (stateValues == null || stateValues == this.AttributeStorage)
+            {
+                return;
+            }
+
+            var attributeStorage = this.AttributeStorage ??= new List<KeyValuePair<string, object?>>(stateValues.Count);
+            attributeStorage.AddRange(stateValues);
```

Comment:
add a comment this is what triggers the iteration/enumeration to "snap" what we need before things could get disposed.

Reply:
Updated